src/System.Text.Json/src/System/Text/Json/JsonReaderState.cs

```diff
@@ -0,0 +1,110 @@
+        /// input ReadOnlySequence&lt;byte&gt;. If the <see cref="JsonUtf8Reader"/> was constructed
+        /// with a ReadOnlySpan&lt;byte&gt; instead, this will always return a default <see cref="SequencePosition"/>.
+        /// </summary>
+        public SequencePosition Position => _sequencePosition;
+
+        /// <summary>
+        /// Returns the total amount of bytes consumed by the <see cref="JsonUtf8Reader"/> so far
+        /// for the given UTF-8 encoded input text.
+        /// </summary>
+        public long BytesConsumed => _bytesConsumed;
+
+        /// <summary>
+        /// Constructs a new <see cref="JsonReaderState"/> instance.
+        /// </summary>
+        /// <param name="maxDepth">Sets the maximum depth allowed when reading JSON, with the default set as 64.
+        /// Reading past this depth will throw a <exception cref="JsonReaderException"/>.</param>
+        /// <param name="commentHandling">Defines how the <see cref="JsonUtf8Reader"/> should handle comments when reading through the JSON.
+        /// By default, the <see cref="JsonUtf8Reader"/> treats comments within the JSON as invalid.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown when the max depth is set to a non-positive value (&lt;= 0)
+        /// </exception>
+        /// <remarks>
+        /// An instance of this state must be passed to the <see cref="JsonUtf8Reader"/> ctor with the JSON data.
+        /// Unlike the <see cref="JsonUtf8Reader"/>, which is a ref struct, the state can survive
+        /// across async/await boundaries and hence this type is required to provide support for reading
+        /// in more data asynchronously before continuing with a new instance of the <see cref="JsonUtf8Reader"/>.
+        /// </remarks>
+        public JsonReaderState(int maxDepth = StackFreeMaxDepth, JsonCommentHandling commentHandling = JsonCommentHandling.Default)
+        {
+            if (maxDepth <= 0)
+                ThrowHelper.ThrowArgumentException_MaxDepthMustBePositive();
+
+            _stackFreeContainer = default;
+            _lineNumber = default;
+            _lineBytePosition = default;
+            _bytesConsumed = default;
+            _currentDepth = default;
+            _maxDepth = maxDepth;
+            _inObject = default;
+            _isNotPrimitive = default;
+            _tokenType = default;
+            _readerOptions = new JsonReaderOptions { CommentHandling = commentHandling };
+            _sequencePosition = default;
+
+            // Only allocate the stack if the user explicitly sets the JsonReaderOptions
+            // by providing a custom JsonCommentHandling OR if the user explicitly sets the
+            // max depth to be larger than 64. This way we avoid allocations in the common, default cases.
+            if (commentHandling == JsonCommentHandling.AllowComments || maxDepth > StackFreeMaxDepth)
+                _stack = new Stack<JsonTokenType>();
```

Comment:
In what situations would someone set a maxDepth > StackFreeMaxDepth?  I'm wondering if in that case it'd be beneficial to pass maxDepth as a capacity to the Stack ctor, or if max depth is just worst case and we generally expect much less than that.

Reply:
Actually, that's probably a bad idea... nevermind.  I could imagine someone saying "I don't want a max depth" and setting it to int.MaxValue.

Reply:
The only reason that I can think of for needing a stack in AllowComments is if it balances `/* this /* is /* nested */ */ */`.   That seems surprising (neither C++ nor C# do); and even if it does, can that be deferred until it is actually needed?

Reply:
The reason why we need stack is not for comment balancing. We need it to remember our previous "valid" json token type before the comment so we know which state to transition to next and whether the state transition is valid.

Reply:
The segments of code I've seen thus far always write to the `ulong` based on `_currentDepth`, and only touch the Stack object once it breaks 64.  So I still don't get why AllowComments + MaxDepth <= 64 would allocate the stack object.

Looking ahead, it looks like the comment-consuming code always pushed onto the stack, but the things that consume from the stack fall back to the `if (_currentDepth <= StackFreeMaxDepth)` logic; so it feels like there's room for misalignment here.

It definitely seems like a pair of helper methods for push/pop would be good; decorated with whatever levels of inlining seems to be warranted.

Reply:
> so it feels like there's room for misalignment here.

There is. Will fix and test.

> It definitely seems like a pair of helper methods for push/pop would be good; decorated with whatever levels of inlining seems to be warranted.

I will see if I can add those. Thanks.